<center> <H1>Stage B Assessment - Winter '23 </H1> </center>

<center> <H4> by </H4> </center>

<center> <H3> Yusuf Okunlola </H3> </center>

Data: Appliances Energy Prediction Dataset - https://drive.google.com/file/d/1Eru_UHVc3WLHVveC9Q8K9QUxlzYeHt18/view?usp=share_link

The dataset for the remainder of this quiz is the Appliances Energy Prediction data. The data set is at 10 min for about 4.5 months. The house temperature and humidity conditions were monitored with a ZigBee wireless sensor network. Each wireless node transmitted the temperature and humidity conditions around 3.3 min. Then, the wireless data was averaged for 10 minutes periods. The energy data was logged every 10 minutes with m-bus energy meters. Weather from the nearest airport weather station (Chievres Airport, Belgium) was downloaded from a public data set from Reliable Prognosis (rp5.ru), and merged together with the experimental data sets using the date and time column. Two random variables have been included in the data set for testing the regression models and to filter out non predictive attributes (parameters). The attribute information can be seen below.

Attribute Information:

Date, time year-month-day hour:minute:second

Appliances, energy use in Wh

lights, energy use of light fixtures in the house in Wh

T1, Temperature in kitchen area, in Celsius

RH_1, Humidity in kitchen area, in %

T2, Temperature in living room area, in Celsius

RH_2, Humidity in living room area, in %

T3, Temperature in laundry room area

RH_3, Humidity in laundry room area, in %

T4, Temperature in office room, in Celsius

RH_4, Humidity in office room, in %

T5, Temperature in bathroom, in Celsius

RH_5, Humidity in bathroom, in %

T6, Temperature outside the building (north side), in Celsius

RH_6, Humidity outside the building (north side), in %

T7, Temperature in ironing room , in Celsius

RH_7, Humidity in ironing room, in %

T8, Temperature in teenager room 2, in Celsius

RH_8, Humidity in teenager room 2, in %

T9, Temperature in parents room, in Celsius

RH_9, Humidity in parents room, in %

To, Temperature outside (from Chievres weather station), in Celsius

Pressure (from Chievres weather station), in mm Hg

RH_out, Humidity outside (from Chievres weather station), in %

Wind speed (from Chievres weather station), in m/s

Visibility (from Chievres weather station), in km

Tdewpoint (from Chievres weather station), °C

rv1, Random variable 1, nondimensional

rv2, Random variable 2, nondimensional

## Questions

1. The percent of the total variation of the dependent variable Y explained by the set of independent variables X is measured by

        Answer: Coefficient of Determination (R2)

2. How do you define a Residual?

        Answer: The difference between the observed value of the dependent variable and the predicted value of the dependent variable

3. The straight line graph of the equation Y = a + BX, the slope is horizontal if

        Answer: b = 0

4. Which of the one is true about Heteroskedasticity?

        Answer: Linear Regression with varying error terms

5. Generally, which of the following method(s) is used for predicting continuous dependent variables?

        1. Linear Regression

        2. Logistic Regression

        Answer: 1 only

6. From the following options below, which of these is/are true about “Ridge” or “Lasso” regression methods in case of feature selection?

        Answer: Lasso regression uses subset selection of features

7. Which of the following sentences is/are true about outliers in Linear Regression:
    
        Answer: Linear regression is not sensitive to outliers

8. Which of the following metrics can be used for evaluating regression models?
        1. R Squared

        2. Adjusted R Squared

        3. F Statistics

        4. RMSE / MSE / MAE
        
        Answer: 1, 2, 3 and 4

9. A best fit line relating X and Y has a R-Squared value of 0.75. How do I interpret this information?

        Answer: 75% of the variance in Y is explained by X

10. Which of the following measures is optimal for comparing the goodness of the fit of competing regression models involving the same dependent variable?

        Answer: Standard deviation of the residuals

11. The Lasso can be interpreted as least-squares linear regression where:
    
        Answer: Weights are regularized with the L1 norm

## Machine Learning Task

In [1]:
# import libraries
import pandas as pd
import numpy as np

#  import preprocessing libraries
from sklearn.model_selection import train_test_split
from sklearn.metrics import  mean_absolute_error, mean_squared_error, r2_score
from sklearn.preprocessing import MinMaxScaler
       
# import regression models
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso

import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

In [2]:
# load dataset
df = pd.read_csv("dataset/energydata_complete.csv")
print("Data imported successfully")

Data imported successfully


In [3]:
# inspect dataframe by printing out the first 5 rows
df.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

In [4]:
# check the shape of the dataset i.e. the number of rows and columns
df.shape

(19735, 29)

In [5]:
# information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19735 entries, 0 to 19734
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         19735 non-null  object 
 1   Appliances   19735 non-null  int64  
 2   lights       19735 non-null  int64  
 3   T1           19735 non-null  float64
 4   RH_1         19735 non-null  float64
 5   T2           19735 non-null  float64
 6   RH_2         19735 non-null  float64
 7   T3           19735 non-null  float64
 8   RH_3         19735 non-null  float64
 9   T4           19735 non-null  float64
 10  RH_4         19735 non-null  float64
 11  T5           19735 non-null  float64
 12  RH_5         19735 non-null  float64
 13  T6           19735 non-null  float64
 14  RH_6         19735 non-null  float64
 15  T7           19735 non-null  float64
 16  RH_7         19735 non-null  float64
 17  T8           19735 non-null  float64
 18  RH_8         19735 non-null  float64
 19  T9  

In [6]:
#  check for the percentage of missingness in each column
a = pd.DataFrame(df.isnull().sum() / len(df) *100, columns=['% of missing values'])
a

% of missing values
date                         0.0
Appliances                   0.0
lights                       0.0
T1                           0.0
RH_1                         0.0
T2                           0.0
RH_2                         0.0
T3                           0.0
RH_3                         0.0
T4                           0.0
RH_4                         0.0
T5                           0.0
RH_5                         0.0
T6                           0.0
RH_6                         0.0
T7                           0.0
RH_7                         0.0
T8                           0.0
RH_8                         0.0
T9                           0.0
RH_9                         0.0
T_out                        0.0
Press_mm_hg                  0.0
RH_out                       0.0
Windspeed                    0.0
Visibility                   0.0
Tdewpoint                    0.0
rv1                          0.0
rv2                          0.0

**Evaluation**

The dataset has 29 columns and 19735 rows. 

The dataset has 25 numerical columns and 4 categorical columns. 

The dataset has 0 missing values.

In [7]:
# descriptive statistics of the dataset
df.describe()

Appliances        lights            T1          RH_1            T2  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      97.694958      3.801875     21.686571     40.259739     20.341219   
std      102.524891      7.935988      1.606066      3.979299      2.192974   
min       10.000000      0.000000     16.790000     27.023333     16.100000   
25%       50.000000      0.000000     20.760000     37.333333     18.790000   
50%       60.000000      0.000000     21.600000     39.656667     20.000000   
75%      100.000000      0.000000     22.600000     43.066667     21.500000   
max     1080.000000     70.000000     26.260000     63.360000     29.856667   

               RH_2            T3          RH_3            T4          RH_4  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      40.420420     22.267611     39.242500     20.855335     39.026904   
std        4.069813      2.006111      3.254576      2.042884      4.341321   
min       20.463333     17.200000     28.766667     15.100000     27.660000   
25%       37.900000     20.790000     36.900000     19.530000     35.530000   
50%       40.500000     22.100000     38.530000     20.666667     38.400000   
75%       43.260000     23.290000     41.760000     22.100000     42.156667   
max       56.026667     29.236000     50.163333     26.200000     51.090000   

       ...            T9          RH_9         T_out   Press_mm_hg  \
count  ...  19735.000000  19735.000000  19735.000000  19735.000000   
mean   ...     19.485828     41.552401      7.411665    755.522602   
std    ...      2.014712      4.151497      5.317409      7.399441   
min    ...     14.890000     29.166667     -5.000000    729.300000   
25%    ...     18.000000     38.500000      3.666667    750.933333   
50%    ...     19.390000     40.900000      6.916667    756.100000   
75%    ...     20.600000     44.338095     10.408333    760.933333   
max    ...     24.500000     53.326667     26.100000    772.300000   

             RH_out     Windspeed    Visibility     Tdewpoint           rv1  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      79.750418      4.039752     38.330834      3.760707     24.988033   
std       14.901088      2.451221     11.794719      4.194648     14.496634   
min       24.000000      0.000000      1.000000     -6.600000      0.005322   
25%       70.333333      2.000000     29.000000      0.900000     12.497889   
50%       83.666667      3.666667     40.000000      3.433333     24.897653   
75%       91.666667      5.500000     40.000000      6.566667     37.583769   
max      100.000000     14.000000     66.000000     15.500000     49.996530   

                rv2  
count  19735.000000  
mean      24.988033  
std       14.496634  
min        0.005322  
25%       12.497889  
50%       24.897653  
75%       37.583769  
max       49.996530  

[8 rows x 28 columns]

# Tasks

From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). 

What is the R^2 value in two d.p.?

In [8]:
# instantiate response and target variables
X = df[["T2"]]
y = df[["T6"]]

# Split data into feature matrix `X_train` and target vector `y_train`.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


# normalize data using MinMaxScaler
ss = MinMaxScaler()
scalled_x_train = ss.fit_transform(X_train)
scalled_x_test = ss.transform(X_test)

# fit model to scaled data
model = LinearRegression().fit(scalled_x_train, y_train)

# predict on scaled test data
y_pred = model.predict(scalled_x_test)

# calculate r2 score
score = r2_score(y_test, y_pred)
print('R-squared:', round(score, 2)) 

R-squared: 0.64


Normalize the dataset using the MinMaxScaler after removing the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a random state of 42 (for reproducibility). Run a multiple linear regression using the training set and evaluate your model on the test set. 

Answer the following questions:

What is the Mean Absolute Error (in two decimal places)?

### **Preparing the data**

In [9]:
# drop date and lights columns
df.drop(columns=["date", "lights"], inplace=True)

In [10]:
# instantiate a minmax scaler for normalisation

scaler = MinMaxScaler()
normalised_df = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)

In [11]:
# The next step is to divide the data into "attributes" (inputs) and "labels" (outputs)
target = "Appliances"
X = normalised_df.drop(columns=[target])
y = normalised_df[target]

In [12]:
# split data into training and test sets with a 70:30 ratio and a random state of 42 (for reproducibility)

x_train, x_test, y_train, y_test = train_test_split(X, y,test_size=0.3, random_state=42)

### **Training the Algorithm**
We have split our data into training and testing sets, and now is finally the time to train our algorithm. 

In [13]:
# instantiate LinearRegression model

linear_model = LinearRegression()

#fit the model to the training dataset
linear_model.fit(x_train, y_train)

#obtain predictions
predicted_values = linear_model.predict(x_test)

In [14]:
# Mean Absolute Error (MAE)

mae = mean_absolute_error(y_test, predicted_values)
print('Mean Absolute Error:', round(mae, 2)) 

Mean Absolute Error: 0.05


In [15]:
# Residual Sum of Squares (RSS)

rss = np.sum(np.square(y_test - predicted_values))
print('Residual Sum of Squares:', round(rss, 2)) 

Residual Sum of Squares: 45.35


In [16]:
# Root Mean Square Error (RMSE)

rmse = np.sqrt(mean_squared_error(y_test, predicted_values))
print('Root Mean Square Error:', round(rmse, 3)) 

Root Mean Square Error: 0.088


In [17]:
# R-Squared i.e. coefficient of determination

r2_score = r2_score(y_test, predicted_values)
round(r2_score, 2)
print('R-squared:', round(r2_score, 2)) 

R-squared: 0.15


Obtain the feature weights from your linear model above. Which features have the lowest and highest weights respectively?

In [18]:
# obtain feature weights for linear regression model

coefficient = linear_model.coef_

feature_weight = pd.DataFrame(coefficient, X.columns, columns=["Coefficient"]).sort_values("Coefficient")
feature_weight

Coefficient
rv2         -8.592401e+10
RH_2        -4.567636e-01
T_out       -3.218442e-01
T2          -2.362284e-01
T9          -1.899250e-01
RH_8        -1.576224e-01
RH_out      -7.764176e-02
RH_7        -4.458982e-02
RH_9        -3.978718e-02
T5          -1.568567e-02
T1          -3.281051e-03
Press_mm_hg  6.860741e-03
T7           1.029952e-02
Visibility   1.229794e-02
RH_5         1.603428e-02
RH_4         2.637866e-02
T4           2.899661e-02
Windspeed    2.919372e-02
RH_6         3.805257e-02
RH_3         9.605492e-02
T8           1.020055e-01
Tdewpoint    1.177459e-01
T6           2.364437e-01
T3           2.906759e-01
RH_1         5.535598e-01
rv1          8.592401e+10

Train a ridge regression model with an alpha value of 0.4. 

Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [19]:
# instantiate ridge model

ridge_reg = Ridge(alpha=0.4)
ridge_reg.fit(x_train, y_train)

#obtain predictions
predicted_values = ridge_reg.predict(x_test)

# Root Mean Square Error (RMSE)

rmse = np.sqrt(mean_squared_error(y_test, predicted_values))
print('Root Mean Square Error:', round(rmse, 3)) 

Root Mean Square Error: 0.088


Answer: There was no change for the RMSE of Linear Regression model and the Ridge Regression model.

Both models have RMSE value of 0.088

Train a lasso regression model with an alpha value of 0.001 and obtain the new feature weights with it. 

How many of the features have non-zero feature weights?

In [20]:
# instantiate lasso model

lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(x_train, y_train)

Lasso(alpha=0.001)

In [21]:
# obtain feature weights for lasso regression
coefficient = lasso_reg.coef_

feature_weight = pd.DataFrame(coefficient, X.columns, columns=["Coefficient"]).sort_values(by="Coefficient", ascending=False)
feature_weight

Coefficient
RH_1            0.017880
Windspeed       0.002912
T1              0.000000
T7             -0.000000
rv1            -0.000000
Tdewpoint       0.000000
Visibility      0.000000
Press_mm_hg    -0.000000
T_out           0.000000
RH_9           -0.000000
T9             -0.000000
T8              0.000000
RH_7           -0.000000
RH_6           -0.000000
T6              0.000000
RH_5            0.000000
T5             -0.000000
RH_4            0.000000
T4             -0.000000
RH_3            0.000000
T3              0.000000
RH_2           -0.000000
T2              0.000000
rv2            -0.000000
RH_8           -0.000110
RH_out         -0.049557

Answer: 4 features have non-zero feature weights

In [22]:
#obtain predictions for lasso regression
predicted_values = lasso_reg.predict(x_test)

# Root Mean Square Error (RMSE)
rmse = np.sqrt(mean_squared_error(y_test, predicted_values))

print('Root Mean Square Error:', round(rmse, 3)) 

Root Mean Square Error: 0.094


In [23]:
#comparing the effects of regularisation

def get_weights_df(model, feat, col_name):
#this function returns the weight of every feature
    weights = pd.Series(model.coef_, feat.columns).sort_values()
    weights_df = pd.DataFrame(weights).reset_index()
    weights_df.columns = ['Features', col_name]
    weights_df[col_name].round(3)
    return weights_df

linear_model_weights = get_weights_df(linear_model, x_train, 'Linear_Model_Weight')
ridge_weights_df = get_weights_df(ridge_reg, x_train, 'Ridge_Weight')
lasso_weights_df = get_weights_df(lasso_reg, x_train, 'Lasso_weight')

final_weights = pd.merge(linear_model_weights, ridge_weights_df, on='Features')
final_weights = pd.merge(final_weights, lasso_weights_df, on='Features')

final_weights

Features  Linear_Model_Weight  Ridge_Weight  Lasso_weight
0           rv2        -8.592401e+10      0.000748     -0.000000
1          RH_2        -4.567636e-01     -0.411071     -0.000000
2         T_out        -3.218442e-01     -0.262172      0.000000
3            T2        -2.362284e-01     -0.201397      0.000000
4            T9        -1.899250e-01     -0.188916     -0.000000
5          RH_8        -1.576224e-01     -0.156830     -0.000110
6        RH_out        -7.764176e-02     -0.054724     -0.049557
7          RH_7        -4.458982e-02     -0.045977     -0.000000
8          RH_9        -3.978718e-02     -0.041367     -0.000000
9            T5        -1.568567e-02     -0.019853     -0.000000
10           T1        -3.281051e-03     -0.018406      0.000000
11  Press_mm_hg         6.860741e-03      0.006584     -0.000000
12           T7         1.029952e-02      0.010098     -0.000000
13   Visibility         1.229794e-02      0.012076      0.000000
14         RH_5         1.603428e-02      0.016152      0.000000
15         RH_4         2.637866e-02      0.024579      0.000000
16           T4         2.899661e-02      0.027384     -0.000000
17    Windspeed         2.919372e-02      0.030268      0.002912
18         RH_6         3.805257e-02      0.035519     -0.000000
19         RH_3         9.605492e-02      0.095135      0.000000
20           T8         1.020055e-01      0.101028      0.000000
21    Tdewpoint         1.177459e-01      0.083128      0.000000
22           T6         2.364437e-01      0.217292      0.000000
23           T3         2.906759e-01      0.288087      0.000000
24         RH_1         5.535598e-01      0.519525      0.017880
25          rv1         8.592401e+10      0.000748     -0.000000